### PR TITLE
Add Memory support

### PR DESCRIPTION
**Pull request by danielbeardsley**
Support retrieving and using the memory information that XDebug records.

Adding "memory" as cost format option was the easiest way to get it done.

### CODE REVIEW

File: library/Preprocessor.php [hunk 1]
```diff
@@ -42,10 +42,11 @@ class Webgrind_Preprocessor
      * @param string $outFile File to write preprocessed data to
      * @return void
      */
-    static function parse($inFile, $outFile)
+    static function parse($inFile, $outFile, $readMemory = false)
     {
+        $costPattern = $readMemory ? "%*d %d" : "%d";
```
Comment (jarstelfox) [inline library/Preprocessor.php:47]: Could you explain the reason for this pattern?
The `%*d` is skipping the number correct?
What number is in the second / third positions?

**Review by andyg0808 — APPROVED** [review]
CR :+1: 
Looks reasonable.

File: library/Preprocessor.php (hunk 1 above)
Comment (danielbeardsley) [reply library/Preprocessor.php:47]: Yes, every "cost line" looks like this:
`4 936 160` == `{line number} {time cost in usec} {memory cost in bytes}`

When we want to read memory, we effectively need to skip the time cost and read the last number (which is normally unread).

File: library/FileHandler.php [hunk 2]
```diff
@@ -160,12 +160,14 @@ public function getTraceList() {
...
-        $prepFile = Webgrind_Config::storageDir().$file.Webgrind_Config::$preprocessedSuffix;
+        $readMemory = $costFormat === 'bytes';
+        $memorySuffix = $readMemory ? ".memory" : "";
+        $prepFile = Webgrind_Config::storageDir().$file.$memorySuffix.Webgrind_Config::$preprocessedSuffix;
         try {
             $r = new Webgrind_Reader($prepFile, $costFormat);
         } catch (Exception $e) {
             // Preprocessed file does not exist or other error
-            Webgrind_Preprocessor::parse(Webgrind_Config::xdebugOutputDir().$file, $prepFile);
+            Webgrind_Preprocessor::parse(Webgrind_Config::xdebugOutputDir().$file, $prepFile, $readMemory);
```
Comment (jarstelfox) [inline library/FileHandler.php:170]: Ah! Cool, you are creating two pre-processed files (one for memory, one for CPU)

**Review by jarstelfox — APPROVED** [review]
CR 🌵

**Comment by kthaler** [comment]
QA 💎 - Webgrind now receives and presents memory information.
deploy_block 🦖 - I'm noticing some serious discrepancies between kcachegrind (qcachegrind for me on M1) and Webgrind. For example:

Screenshot 2023-11-03 at 1 53 22 PM

If this is alright, then please disregard the deploy block.

**Comment by danielbeardsley** [comment]
I made a minimal example and the data is coming out correct in both tools for me. Actually, it appears that `kcachegrind` has a bug.

```php
<?

function main() {
   $x[] = useMemory(100);
   $x[] = useMemory(1000);
   $x[] = useMemory(10000);
   $x[] = useMemory(100000);
   $x[] = useMemory(1000000);
   $x[] = useMemory(10000000);
}

function useMemory(int $bytes) {
   return str_repeat("x", $bytes);
}

main();
```

Note that `kcachegrind` says that `main()` has `0` bytes of inclusive cost. This is obviously false, so I'm not sure what's going on. Maybe it's cause I named in `main` like the app entrypoint? I'll try a more complex example.

image

**Comment by danielbeardsley** [comment]
Nah, it seems like it's a bug in `kcachegrind`

Here's a slightly more complex example:

```php
function main() {
   $x[] = use10KB();
   $x[] = use1MB();
}

function use1MB() {
   $overhead = useMemory(100);
   return useMemory(10000000);
}

function use10KB() {
   $overhead = useMemory(100);
   return useMemory(10000);
}

function useMemory(int $bytes) {
   return str_repeat("x", $bytes);
}

main();
```

image

**Comment by danielbeardsley** [comment]
@jarstelfox recommended I change "main" to something else. No dice, `kcachegrind` still shows the bug:

image

**Comment by danielbeardsley** [comment]
dev_block :+1: the binary pre-processor is producing the wrong results.

**Comment by jarstelfox** [comment]
I do wish there was some test suite here.
You should open this up on the upstream 💙

~~Welp, your examples give me confidence that this is working as expected.~~

looks like you found a bug

**Comment by danielbeardsley** [comment]
un_dev_block :+1: The outputs are the same, it was a sorting type auto-detection issue.

**Review by jarstelfox — APPROVED** [review]
CR 🌵

**Review by andyg0808 — APPROVED** [review]
CR :+1:

**Comment by kthaler** [comment]
QA 💎  - The bytes of memory cost option in webgrind is displaying correct values.

un_deploy_block 🦕  - Fixed the problem with my qcachegrind and it is now showing the same/similar values as webgrind.

**Pull request merged**